File: src/c/generics.c
// 4️⃣ Generics and Monomorphization vs C Macros
// 
// Compile: gcc -O2 -S src/c/generics.c -o target/generics.s
// Assembly file: target/generics.s
//
// Key assembly differences to observe:
// - C macros are simple textual substitution
// - Assembly shows inlined multiplication operations
// - No function call overhead (due to macro expansion)
// - May optimize directly to immediate value calculations
// - No type safety checking, relies entirely on preprocessor
// - Potential for unexpected behavior with complex expressions
// - No compile-time type validation

#include <stdio.h>

#define SQUARE(x) ((x)*(x))

int main() {
    printf("%d\n", SQUARE(3));    // Preprocessor expands to ((3)*(3))
    printf("%f\n", SQUARE(2.0));  // Preprocessor expands to ((2.0)*(2.0))
    return 0;
} 

File: src/rust/generics.rs
// 4️⃣ Generics and Monomorphization vs C Macros
// 
// Compile: cargo rustc --bin generics --release -- --emit=asm
// Assembly file: target/release/deps/generics-*.s
//
// Key assembly differences to observe:
// - Rust generates specialized function versions for each type (monomorphization)
// - Assembly shows multiple square function versions: one for i32, one for f64
// - Function names include type information mangling
// - Each version is optimized for its specific type
// - No runtime type checking overhead
// - Zero-cost abstraction with compile-time specialization
// - Type safety guaranteed at compile time

fn square<T: std::ops::Mul<Output = T> + Copy>(x: T) -> T {
    x * x
}

fn main() {
    println!("{}", square(3));    // Generates square::<i32> version
    println!("{}", square(2.0));  // Generates square::<f64> version
} 